### DISCUSSION TITLE
Noob questions :)

### DISCUSSION THREAD

**Discussion by kairosdojo** [boris-kz/CogAlg#84] [Q&A] [answered]
Hello, everyone!

I've stumbled upon this project after reading some comments of Boris in Numenta's discussion site. I'm very intrigued as his it's an approach different from both deep learning / fitting of polynomial functions (which I deem to be filled with issues too) and neuromorphic algorithms (like SNNs or Numenta's / Corical's HTM; which instead I'm more interested in).

I've decided to spend a bit of time going more in depth through your codebase and your Wiki. I've given a first, superficial, reading of the latter, yet I've several questions (and many more might come). I'm sorry anyway - these are pretty newbie doubts (I've many and they are not linked to each other: bear with me :) ). So:

1. About [line_Ps.py](https://github.com/boris-kz/CogAlg/blob/master/line_1D_alg/line_Ps.py) (and the similar implementations in 2D/3D): what is the time complexity of it? It sounds to me it can reach easily O(n²) or even O(n³). Am I right?

2. The idea of building several *fingerprints* of differences (and having a hierarchy of those) might probably benefit from a lookup table / hashmap to speed up computation (for instance in 2D+ inputs). Incidentally, this reminds me (I know, it's a big leap / a stretch) of Jungian  archetypes (but in a less evanescent and more practical way). Is this something you guys envision too (I mean: the possibility to make a lookup table as a base of knowledge created with this system)?

3. I know this has already been asked but it seems pretty crucial to me and a bit neglected: is there a defined roadmap? A kanban board, a trello, wekan, etc? 

4. I've read in [cognitivealgorithm.info](www.cognitivealgorithm.info) this sentence: *And so on, higher levels should be added recursively. Such process is very complex and deeply structured, there is no way it could evolve naturally.* You mean that "it can't arise unless we program it this way explicitly" but, and this is just a curiosity - do you also intend that even biological intelligence did not evolve naturally? (Or do you consider your approach with no relation to biological intelligence, hence that sentence applies only to the former?).

5. The wiki is quite a read: is there by any chance a video, a podcast, some interactive Q&A session you/you guys  partook in or will partake in anytime soon? That would be of tremendous help.

Again, sorry for the noob questions: I'm still trying to wrap my head around these ideas.

**Comment by boris-kz** [comment, marked as answer]
Appreciate your interest Raffaele!

1: The process is selective per level of an input pattern, so time and cached memory depends on predictive value of that pattern, only 1st-level comparison is default.
2: The "archetypes" should be learned, so that's no different from patterns, and the differences you mentioned are formed by their cross-comp: my core process.
3: The wiki? It's not very detailed as a roadmap, but this is basic research at this point, we figure things out as we go.
4: Yes, I meant my comparison-first scheme, neuromorphic intelligence obviously did evolve on its own. I see my scheme as the next level, it can't emerge without NI just like NI can't emerge without life. Rationally designed substrate-independent intelligence is intrinsically different.
5: I try integrate answers to frequent question into readme and wiki, QAs lack structure and are far more difficult to wade through. But we can always talk, skype or zoom? I am boris.kazachenko on skype, best time is early mornings EST.

Good questions, great to see some serious interest!

**Reply by kairosdojo** [reply]
Ciao Boris,

5: that would a pleasure! Only problem is: as I use linux, skype is a bit buggy for me. I could pass you a zoom link via skype, though, right before our call. What do you think about tomorrow (or another day if you can't) at 8.00 am EST (which should correspond to 1pm BST, my time)? Is it too early / late? 

Have a nice... morning, I guess :)

**Reply by boris-kz** [reply]
Sounds good, your 1PM is perfect. And of course we can continue here if you think of something specific.

**Reply by kairosdojo** [reply]
Perfect, see you tomorrow then!
